### PR TITLE
Introducing a password handler authored by @DASPRiD 

### PR DESCRIPTION
**Pull request by awartoft**
Whats left be done
- [ ] What key to use for the `Zend\Crypt\Password\HandlerManager` 
- [ ] Integrate the HandlerManager into the mvc service listener factory
- [ ] Documentation

### CODE REVIEW

File: library/Zend/Crypt/Password/Algorithm/Bcrypt.php [hunk 1]
```diff
@@ -0,0 +1,103 @@
...
+        }
+
+        return $this->backend;
+    }
+
+    /**
+     * @param  BcryptHasher $backend
+     * @return Bcrypt
+     */
+    public function setBackend(BcryptHasher $backend)
```
Comment (ojhaujjwal) [inline library/Zend/Crypt/Password/Algorithm/Bcrypt.php]: IMO Setter is not needed. Constructor will do!

**Comment by ojhaujjwal** [comment]
I like the overall idea. :+1:

File: library/Zend/Crypt/Password/HandlerInterface.php [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+     * @param  string $password
+     * @param  string $hash
+     * @return boolean
+     */
+    public function compare($password, $hash);
+
+    /**
+     * Determines whether the hash should be re-hashed.
+     *
+     * This is used primarily by the authentication adapter, which will verify
```
Comment (DASPRiD) [inline library/Zend/Crypt/Password/HandlerInterface.php]: This comment is pretty much tied to BaconUser; should be updated.

**Comment by padraic** [comment]
Note to monitor https://github.com/zendframework/zf2/pull/7141 which is effectively proposing a new default (a minor twist on plain bcrypt to preserve entropy from >72 byte passwords).

**Comment by weierophinney** [comment]
@macnibblet What remains to be done? Can this be ready for 2.4?

**Comment by macnibblet** [comment]
@weierophinney Everything that is listed in the task list, if i don't manage to look into it on sunday i'm doubtful that I will have time to finish it before the 2.4 release.

**Comment by weierophinney** [comment]
@macnibblet I'll give you until the end of the first week of March to get code ready. :)

**Comment by malukenho** [comment]
@macnibblet just to remember /cc @weierophinney

**Comment by moderndeveloperllc** [comment]
@macnibblet I'm concerned that the MCF identifiers that you are using for SHA1 and MD5 are not the standard identifiers that are used in the standard Modular Crypt Format. `$1$` is standard for MD5. SHA1 is not 'standard' (mainly because it's not strong enough), but I have seen a few implementations that use `$sha1`. At least adding these to the checks with what you have will help developers who have to integrate with other languages that use the standard C `crypt()` library.

**Comment by DASPRiD** [comment]
@moderndeveloperllc do you have a reference for the $1$ marker? Apart from that I'd agree to use those instead. Also we may want to remove the "simple-" part from the classes. No clue why I had chosen those names.

**Comment by moderndeveloperllc** [comment]
@DASPRiD The `$1$` is from the [`crypt()` library docs](https://www.freebsd.org/cgi/man.cgi?query=crypt(3)). It also lists `$5$` for SHA256 and `$6$` for SHA512 along with the $2$ series for bcrypt.

I might actually take back what I've said in my previous comment, because `$1$` usually refers to an md5 where you are storing the salt - kinda like: `$1$fauihdfaoihfad$caeioanceoi039rasi09`. The code in these files are looking for a plain - single round MD5 with nothing but the hashed value itself stored. In the real world, would these even have an MCF?

FYI, The best explanation/reference I found was with Python's [PassLib documentation](http://pythonhosted.org/passlib/modular_crypt_format.html). As far as I can tell, there is no "standard" in that there are no IEEE or RFC standards out there. Most links I've seen refer to the PassLib docs. For instance, in the [scrypt module I built](https://bitbucket.org/moderndeveloperllc/zf2-scrypt), I decided to use `$s0$` because the leading [Java scrypt library](https://github.com/wg/scrypt) used that.

**Comment by GeeH** [comment]
This issue has been moved from the `zendframework` repository as part of the bug migration program as outlined here - http://framework.zend.com/blog/2016-04-11-issue-closures.html
New issue can be found at: https://github.com/zendframework/zend-crypt/issues/30

**Pull request closed without merging**